docs/api.rst

```diff
@@ -1,23 +1,203 @@
 General API customization
 =========================
 
-.. confval:: include_object_description_fields_in_toc
+This theme supports a number of options that can be customized for each
+domain/object type pair.
 
-   :python:`bool` indicating whether to include domain object description
-   fields, like "Parameters", "Returns", "Raises", etc. in the table of
-   contents.  Defaults to :python:`True`.
+.. confval:: object_description_options_map
+
+   :py:obj:`dict` mapping strings of the form :python:`"domain:objtype"` to a
+   dictionary of supported `object-description-options`.
+
+   This option may be used to specify options for a single option type.  To
```

Comment:
I'm partial to just removing this config option entirely in favor of only using `object_description_options_patterns`. This would make our lives a bit simpler, yes?

What is stopping
```py
object_description_options_patterns = [
    ("cpp:functionParam", dict(include_in_toc=False, generate_synopses=None)),
    ("cpp:function", dict(include_fields_in_toc=False, generate_synopses=None)),
]
```
from being equivalent to 
```py
object_description_options_map = {
    "cpp:functionParam": dict(include_in_toc=False, generate_synopses=None),
    "cpp:function": dict(include_fields_in_toc=False, generate_synopses=None),
}
```

It looks like regex is only applied to the `object_description_options_patterns` in which case this needs to be altered.
```suggestion
   This option may be used to specify options for a single object type.  To
```

Reply:
Done.  Originally I added this for efficiency, to avoid having to scan through all patterns for all object types, given that there are a lot of rules for `toc_icon_class`.  The options are cached per object type, so it wouldn't be too bad, but I am still wary of adding any quadratic complexity to the build.

However I realized I could still apply the same optimization even without needing a separate map.